walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,85 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.stream.Stream;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration())
+      return new ArrayList<TimeRange>();
+    List<TimeRange> freeTimes = queryWithOptionalAttendees(events, request);
+    if (freeTimes.size() == 0 && request.getAttendees().size() > 0)
+      freeTimes = queryWithOnlyMandatoryAttendees(events, request);
+    return freeTimes;
+   
+  }
+
+  private List<TimeRange> queryWithOptionalAttendees(Collection<Event> events, MeetingRequest request) {
+    Set<String> allAttendees = new HashSet<String>() {{
+      addAll(request.getAttendees());
+      addAll(request.getOptionalAttendees());
+    }};
+    List<TimeRange> busyTimes = getBusyTimes(events, allAttendees);
+    return getFreeTimes(busyTimes, request.getDuration());
+  }
+
+  private List<TimeRange> queryWithOnlyMandatoryAttendees(Collection<Event> events, MeetingRequest request) {
+    List<TimeRange> busyTimes = getBusyTimes(events, request.getAttendees());
+    return getFreeTimes(busyTimes, request.getDuration());
+  }
+
+  private List<TimeRange> getBusyTimes(Collection<Event> events, Collection<String> requestAttendees) {
+    List<TimeRange> busyTimes = new ArrayList<TimeRange>();
+    busyTimes = 
+      events.stream()
+      .filter((event) -> containsAttendee(event, requestAttendees))
+      .map((event) -> event.getWhen())
+      .sorted(TimeRange.ORDER_BY_START)
+      .collect(Collectors.toList());
+    return busyTimes;
+  }
+
+  private List<TimeRange> getFreeTimes(List<TimeRange> busyTimes, long duration) {
+    List<TimeRange> freeTimes = new ArrayList<TimeRange>();
+    if(busyTimes.size() == 0) {
+      freeTimes.add(TimeRange.WHOLE_DAY);
+    } else {
+      int index = 0;
+      TimeRange first = busyTimes.get(index);
+      if (first.start() - TimeRange.START_OF_DAY >= duration)
+        freeTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, first.start(),false));
+      int start = first.end();
+      while(index < busyTimes.size() - 1) {
+        first = busyTimes.get(index);
+        TimeRange second = busyTimes.get(index + 1);
+        if (first.overlaps(second)) {
```

Comment:
Since first.start <= second.start, if they overlap, then first.end < second.end should always be true? Also start < second.end should always be true?

Reply:
Oooo... Forget this comment. I didn't notice that `overlaps` also contains `contains`.